pkg/tokenvalidation/timeoutvalidator.go

```diff
@@ -0,0 +1,233 @@
+package oauth
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/klog"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	userv1 "github.com/openshift/api/user/v1"
+	oauthclient "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	oauthclientlister "github.com/openshift/client-go/oauth/listers/oauth/v1"
+	"k8s.io/kubernetes/openshift-kube-apiserver/authentication/oauth/rankedset"
+)
+
+var errTimedout = errors.New("token timed out")
+
+// Implements rankedset.Item
+var _ = rankedset.Item(&tokenData{})
+
+type tokenData struct {
+	token *oauthv1.OAuthAccessToken
+	seen  time.Time
+}
+
+func (a *tokenData) timeout() time.Time {
+	return a.token.CreationTimestamp.Time.Add(time.Duration(a.token.InactivityTimeoutSeconds) * time.Second)
+}
+
+func (a *tokenData) Key() string {
+	return a.token.Name
+}
+
+func (a *tokenData) Rank() int64 {
+	return a.timeout().Unix()
+}
+
+func timeoutAsDuration(timeout int32) time.Duration {
+	return time.Duration(timeout) * time.Second
+}
+
+type TimeoutValidator struct {
+	oauthClients   oauthclientlister.OAuthClientLister
+	tokens         oauthclient.OAuthAccessTokenInterface
+	tokenChannel   chan *tokenData
+	data           *rankedset.RankedSet
+	defaultTimeout time.Duration
+	tickerInterval time.Duration
+
+	// fields that are used to have a deterministic order of events in unit tests
+	flushHandler    func(flushHorizon time.Time) // allows us to decorate this func during unit tests
+	putTokenHandler func(td *tokenData)          // allows us to decorate this func during unit tests
+	clock           clock.Clock                  // allows us to control time during unit tests
+}
+
+func NewTimeoutValidator(tokens oauthclient.OAuthAccessTokenInterface, oauthClients oauthclientlister.OAuthClientLister, defaultTimeout int32, minValidTimeout int32) *TimeoutValidator {
```

Comment:
is there a hard limit after which a token will timeout?

Reply:
last request time + whatever timeout is configured